lib/ecmascript.ts

```diff
@@ -3679,151 +3831,204 @@ export const ES = ObjectAssign({}, ES2020, {
-      case 'microsecond':
-        return ES.BalanceTime(hour, minute, second, millisecond, result, 0);
-      case 'nanosecond':
-        return ES.BalanceTime(hour, minute, second, millisecond, microsecond, result);
-    }
-  },
-  DaysUntil: (earlier, later) => {
-    return ES.DifferenceISODate(
-      GetSlot(earlier, ISO_YEAR),
-      GetSlot(earlier, ISO_MONTH),
-      GetSlot(earlier, ISO_DAY),
-      GetSlot(later, ISO_YEAR),
-      GetSlot(later, ISO_MONTH),
-      GetSlot(later, ISO_DAY),
-      'day'
-    ).days;
-  },
-  MoveRelativeDate: (calendar, relativeTo, duration) => {
-    const options = ObjectCreate(null);
-    const later = ES.CalendarDateAdd(calendar, relativeTo, duration, options);
-    const days = ES.DaysUntil(relativeTo, later);
-    relativeTo = ES.CreateTemporalDateTime(
-      GetSlot(later, ISO_YEAR),
-      GetSlot(later, ISO_MONTH),
-      GetSlot(later, ISO_DAY),
-      GetSlot(relativeTo, ISO_HOUR),
-      GetSlot(relativeTo, ISO_MINUTE),
-      GetSlot(relativeTo, ISO_SECOND),
-      GetSlot(relativeTo, ISO_MILLISECOND),
-      GetSlot(relativeTo, ISO_MICROSECOND),
-      GetSlot(relativeTo, ISO_NANOSECOND),
-      GetSlot(relativeTo, CALENDAR)
-    );
-    return { relativeTo, days };
-  },
-  MoveRelativeZonedDateTime: (relativeTo, years, months, weeks, days) => {
-    const timeZone = GetSlot(relativeTo, TIME_ZONE);
-    const calendar = GetSlot(relativeTo, CALENDAR);
-    const intermediateNs = ES.AddZonedDateTime(
-      GetSlot(relativeTo, INSTANT),
-      timeZone,
-      calendar,
-      years,
-      months,
-      weeks,
-      days,
-      0,
-      0,
-      0,
-      0,
-      0,
-      0
-    );
-    return ES.CreateTemporalZonedDateTime(intermediateNs, timeZone, calendar);
-  },
-  AdjustRoundedDurationDays: (
+    dayLengthNs
+  ));
+  ({ year, month, day } = BalanceISODate(year, month, day + deltaDays));
+  return { year, month, day, hour, minute, second, millisecond, microsecond, nanosecond };
+}
+
+export function RoundTime(
+  hour,
+  minute,
+  second,
+  millisecond,
+  microsecond,
+  nanosecond,
+  increment,
+  unit,
+  roundingMode,
+  dayLengthNs = DAY_NANOS
+) {
+  let quantity = bigInt.zero;
+  switch (unit) {
+    case 'day':
+    case 'hour':
+      quantity = bigInt(hour);
+    // fall through
+    case 'minute':
+      quantity = quantity.multiply(60).plus(minute);
+    // fall through
+    case 'second':
+      quantity = quantity.multiply(60).plus(second);
+    // fall through
+    case 'millisecond':
+      quantity = quantity.multiply(1000).plus(millisecond);
+    // fall through
+    case 'microsecond':
+      quantity = quantity.multiply(1000).plus(microsecond);
+    // fall through
+    case 'nanosecond':
+      quantity = quantity.multiply(1000).plus(nanosecond);
+  }
+  const nsPerUnit = unit === 'day' ? dayLengthNs : bigInt(nsPerTimeUnit[unit]);
+  const rounded = RoundNumberToIncrement(quantity, nsPerUnit.multiply(increment), roundingMode);
```

Comment:
It seems that `RoundNumberToIncrement` method expects increment to be a Number, not a bigint.

Reply:
Looking upstream from this code, I think the root-cause bug in the current `main` (before this PR) is that the implementation of `round()` in zoneddatetime.ts passes a bigint for `dayLengthNs`. AFAICT, in current main this is the only case that leads to a bigint being passed as `increment`. 

For the purposes of this PR, I suspect that the easiest path would be to change `ZonedDateTime.p.round` to convert `dayLengthNs` to Number before passing to RoundISODateTime. Then (assuming there isn't other use of bigint in these values) all downstream code (including the lines above this comment) could use Number for `dayLengthNs` and `increment` parameters, and we'd only have to JIT-coerce them to bigint before combining them in arithmetic operations with bigint values.

I think it's reasonable to discuss whether we'd want to convert these and similar parameters to bigint, but that seems out of scope to this PR, and might have a perf or bundle-size impact so IMHO seems worth considering separately after this PR is merged.